tests/unit/test_hnsw.cpp

```diff
@@ -1630,6 +1633,130 @@ TYPED_TEST(HNSWTest, testInitialSizeEstimation_No_InitialCapacity) {
     VecSimIndex_Free(index);
 }
 
+TYPED_TEST(HNSWTest, testIncomingEdgesSize) {
+    size_t dim = 4;
+    size_t n = DEFAULT_BLOCK_SIZE;
+    size_t bs = DEFAULT_BLOCK_SIZE;
+    size_t efC = n;
+
+    // In this test we add identical vectors.
+    // Expected results:
+    // in level 0: each new node is connected to the M last inserted nodes, and there will be no
+    // unidirectional edges. The reason for that is that each node has room for 2 * M neighbors, but
+    // only connects the M vectors whose ids are the maximal (due to the secondary sorting by id in
```

Comment:
Are you sure the vectors in level 0 connect only to `M` vectors? They can get up to 2M. As far as I understand, the flow is the same in each level, the only relaxtion for level 0 is that you are allowed to add up to addtional M connections without the need to "swap" outgoing edges from your earliest neighbour to the newest. This means that your observation about the increasing amount of outgoing edges for `curr_idx_at_level < M` and decreasing amount for `curr_reverse_idx_at_level < M` holds for the total number of edges (out+in) for each node in each level (meaning the other nodes will have 2M connections overall)